### ISSUE TITLE
Keep button to Play Edited Scene besides Reload Playing Scene (current F6) when game is running (as in Godot 3)

### ISSUE THREAD

**Issue by hsandt** [godotengine/godot-proposals#6550]
### Current state and version comparison

**UPDATE 2024-02-28**
I added this section to track progress of the editor behavior.

**v4.0.2.stable.official [7a0977ce2]**

The editor behaves as described in the original post:
- when nothing is running, F6 runs the opened scene
- when the game has been run once via F6 (opened scene), F6 reruns the same scene again independently from currently opened scene

**v4.1.1.stable.official [bd6af8e0e]**
**v4.2.1.stable.official [b09f793f5]**

Due to a **bug**, the editor almost behaves as I suggested, except there is no button to force rerun the same scene for retrocompatibility, and the buttons didn't change appearance nor tooltip (causing some confusion):
- when nothing is running, F6 runs the opened scene
- when the game has been run once via F6 (opened scene), F6 runs the currently opened scene independently from the scene previously run

Since this happened due to some bug, we shouldn't count on it. In fact, the fix PR https://github.com/godotengine/godot/pull/87170 has already been approved for 4.3.

**v4.3**

... (behavior probably back to 4.0)

### Describe the project you are working on

A platformer with multiple scenes

### Describe the problem or limitation you are having in your project

If the game is running from a certain scene, and I open a different one then Run Current Scene with F6, because the button was transformed into the special "Reload the played scene" button (shortcut is the same, but icon and tooltip changed), it will reload the game from that same old scene instead of the new one I just opened.

In practice, it happens when I open a bad scene (e.g. the Player.tscn which only contains the player and cannot work alone) and I encounter an error that sends me to the code editor. I realize I ran the wrong scene, so I open my Level.tscn and re-run with F6. Except the new behavior or Run Current Scene will replay the bad scene again and again.

The new behavior was added in https://github.com/godotengine/godot/issues/31515 and there was a doubt about confusing users at the time. This issue is exactly about that confusion.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

We could add a third button that basically has the old Godot 3 behavior: it always runs the scene currently opened in the editor.

To avoid redundancy, when not playing, we would only show the button to Play Edited Scene.
When playing, we show both the button to Play Edited Scene and the existing one to Reload Playing Scene.

In this case, we can either decide
a. to keep the contextual action "Play Edited Scene OR Reload Playing Scene" as we currently have, or
b. to remove redundancy completely but only having a "Reload Playing Scene" button that only works when playing.

If we choose b., users who like the new behavior will need to remember and use two mappings depending on the situation. So I suppose a. is not bad despite the redundancy. I'll assume we choose a. for the mappings below.

For the default mapping we could have:
- F6: always Play Edited Scene, then a less trivial shortcut like Ctrl+F6: Play Edited Scene OR Reload Playing Scene (if playing)
or the reverse:
- F6: Play Edited Scene OR Reload Playing Scene (if playing), then a less trivial shortcut like Ctrl+F6: always Play Edited Scene

Really, it depends on what most users prefer, and may require a survey. If users are good with the new behavior, we can keep F6 for the contextual action and add Ctrl+F6 for the other one. Then users who prefer the old behavior can just reverse the mapping in their preferences.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When not playing, we would have:

![Buttons not playing](https://user-images.githubusercontent.com/1067772/226691693-9060646b-32b6-4feb-8c12-564144329105.png)

And when playing:

![Buttons when playing](https://user-images.githubusercontent.com/1067772/226691698-82c975e1-a01b-424c-a59f-f1bb7900551c.png)

(note how the Play edited scene button is still there)

(mapping tooltip depends on what we choose as default mappings)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The point is to reuse it often anyway. For editor plugins, see below.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think editor plugins can modify the basic buttons. We could make an extra window with Run buttons but that would be overkill to just restore an old functionality. And this enhancement

**Comment by h0lley** [comment]
I was wondering what was wrong for a couple of months now, as in my experience (someone used to the Godot 3 behavior), the F6 function seemed to just randomly launch unexpected scenes.

I spent so much time confusedly switching back and forth between scene tabs, clicking into the editor viewport and the node tree to desperately tell Godot "hey, this is the scene I am currently editing and the one I want you to launch when I press F6". as you can tell I had a really bad time due to that change.

I never click the buttons but always use F5/F6/F8 keys, so I never noticed the "Play current scene" button changing into a reload button, and even if I did, the tooltip does not at all make the new behavior clear.

once I finally realized that the F6 behavior now differs depending on whether test play is still running or not, I reported it as a bug, more thoughts in there: https://github.com/godotengine/godot/issues/75998

so yea while I don't really see the use case for the new reload behavior (you could do the same thing with the old behavior while having greater control and not having to quit when wanting to test another scene), if not reverted I am for adding a second button or a ctrl modifier to bring the original behavior back.

in the case of a modifier, I'd be for the original behavior being plain F6 and the new behavior requiring the modifier as it seems to be more of an advanced use case.

**Comment by AThousandShips** [comment]
> the user this was being changed for

This is intellectually dishonest and claims that this was done on a whim for *one* user rather than being both supported by several others but also the reviewing process

Also you are misrepresenting or misunderstanding their quote, as pointed out in https://github.com/godotengine/godot/issues/75998#issuecomment-1506399989

**Comment by h0lley** [comment]
> supported by several others

it does have a handful of thumbs up, true, my bad. I was only looking at actual participation in the form comments, where - other than core contributors - just one user participated who did voice support for the original behavior, saying that is currently works as they would expect.

> that user didn't raise any issue over the change

pretty save assumption they have long since stopped using Godot - it's three years between issues reported and "fixed".

> but also the reviewing process

where this exact issue has been foreseen by Akien but not further discussed:

> That being said, I can foresee users being equally confused if pressing the "Play Current Scene" button/hotkey would replay the currently playing scene instead of the one they chose to edit.

**Comment by h0lley** [comment]
> Actually that one user asked for the same change as the OP. They expected the button to reload the currently running scene, not the currently edited scene. They expected this to be true, but it wasn't true, hence the request for changes.

for reference, the comment in question:

> I would expect the icon to indicate intention to "reload currently running scene". It sounds like this is the current behavior.
Am I understanding you correctly in that you wish for the updated behavior to be "close all running scenes and reload the first previously opened scene"?

guess everybody just interprets these comments like they want, including myself, of course.

I am not sure how they mean their first sentence, but what is clear to me is that they thought the current behavior (original back then) made sense to them.

they then inquired if OP is suggesting the function to "close all running scenes and reload the first previously opened scene", which is apparently different from what they'd expect but is exactly what ended up being implemented.

**Comment by YuriSizov** [comment]
> I am not sure how they mean their first sentence

Well, what they are saying is they think that the button should reload the currently running scene (and by that everyone means the scene that started the debug session, not the current state of your scene tree in the runtime). And they assume that's what it does. And they assume that the OP asks for something different. Then OP clarifies that, no, this is not what the button does, but it is what they think it should do as well. Both of them are on the same page and request the same thing. It's just that the other commenter assumed this was already the case, when it was not.

**Comment by YuriSizov** [comment]
@h0lley When you press a "Play" button and it changes into a "Restart" button, it is natural and intuitive to expect that clicking it would do _exactly the same_ as clicking "Play" the first time did. There is no extra insight or hidden concepts. This is why the change was made. Because users expected to see exactly the same outcome from pressing the "Restart" button as they saw from pressing the corresponding "Play" button. It's that simple.

But I do agree, that this is not intuitive when pressing the hotkey, and we should probably change that.

**Comment by Deozaan** [comment]
> When you press a "Play" button and it changes into a "Restart" button, it is natural and intuitive to expect that clicking the "Restart" button would do exactly the same as clicking the "Play" button the first time did. 

Yes. The thing the play button did when I clicked it the first time was to run the scene I had currently selected in the editor when I clicked it. When I press it again, I still expect it to run the scene I have currently selected in the editor when I click it, even if that is a different scene compared to last time.

**Comment by YuriSizov** [comment]
That doesn't really fit the definition of doing exactly the same things, nor the common meaning of "Restart", nor the connotation of the icon that we use. And you can still achieve that by stopping the debug session and starting a new one.

**Comment by h0lley** [comment]
ah yes, isn't it fun assuming authority over definitions in an argument.

that circle arrow button could just as well be understood as "repeat".
repeating the action of launching the current main scene.
repeating the action of launching the scene currently being edited.

so really, this is arguing semantics and shouldn't stand in the way of improving UX.

> And you can still achieve that by stopping the debug session and starting a new one.

quoting myself from the other thread: sure, I mentioned this same workaround myself several times in this thread, but as also explained, this is nothing but a needless extra step that each time it is missed may result in an unintended scene starting.

> And I'll say it again: this doesn't apply to the hotkey

I don't think it's a good idea to disconnect the hotkeys from the UI. may just introduce another source of confusion. but if that's what it takes then I'd prefer it over the current state. dedicated restart button as proposed seems to me a much cleaner solution though.

**Comment by Deozaan** [comment]
> That doesn't really fit the definition of doing exactly the same things, nor the common meaning of "Restart", nor the connotation of the icon that we use.

I disagree. I see it as (re)running a function where the current/edited scene is passed in as an argument. The function still does the same thing, but of course with a different argument it will output different results. It's very disorienting to think you've passed in a new argument but see results matching a previous argument.

That's like making changes to your code, running the project, seeing the old behavior and being totally confused about why your new code didn't take any effect, possibly wasting some time debugging or making other code changes, and then realizing the problem all along was that you forgot to save your changes before you ran the project. Except in this case it's not a silly user error, it's the editor itself doing something new/different/unexpected.

This isn't a perfect analogy, but when I press the equal sign on my calculator multiple times in a row, I don't expect it to do "exactly the same thing" as you define it, which would be to evaluate the exact same expression as it did previously. I expect it to keep doing the thing I told it to do, which is to modify some (dynamic) value by another value, which gives me a different result each time I press it. How is anyone supposed to know what the dynamic value is if it keeps changing? The answer is simple and intuitive: it's the value right in front of me at the time when I press the button.

![Calculator](https://user-images.githubusercontent.com/227734/232121290-315e9632-f143-4fad-b18d-398803f0016a.gif)


> And I'll say it again: this doesn't apply to the hotkey, which should probably always launch the current edited scene. (Which you can also do from the right-click menu on the scene tab, btw).

Thank you for saying this again. I had missed it. While I still disagree with your definition of what it means to do "exactly the same thing," I am perfectly happy with allowing the hotkey to do what I expect it to do, even if the button does something I don't expect it to do, because I pretty much never press the button.

That said, I think that since more advanced users tend to use hotkeys, the kind of people who will use the button instead of the hotkey are probably more likely to be novices. And they might be confused about why their test button has been replaced with a new button that does something different than the button that was there before. So I agree that having a separate, dedicated "Restart" button that only appears when running the debug instance is the ideal solution. Especially if it can be assigned its own hotkey, for the more advanced users who would like to take advantage of that behavior.

**Comment by hsandt** [comment]
So I just tested 4.2.1 today and F6 now behaves as I wanted, playing the edited scene no matter what. But the icon/tooltip didn't change, and weirder, I spotted the change to happen between 4.0 and 4.1 (more details in my OP which I updated with a new section at the beginning).

How is it possible? The PR mentioned just above has been decided for 4.3 and cherry picks for 4.1 and 4.2 have been canceled! Did someone change the code and not mention this issue? I checked the [history of `editor_run_bar.cpp`](https://github.com/AThousandShips/godot/commits/595c6248a3e7d988e34d53635737e906c1eed819/editor/gui/editor_run_bar.cpp) just in case but I see nothing around 4.1 release date... Can someone solve this mystery?

**Comment by AThousandShips** [comment]
If you look at the comments you'll see the change that was fixed was an accident in 4.1, not intended, but was not cherry picked because it was considered controversial

**Comment by hsandt** [comment]
> If you look at the comments you'll see the change that was fixed was an accident in 4.1, not intended, but was not cherry picked because it was considered controversial, so the current behavior in 4.2 is not intended, and the one in 4.3 is

Yes I saw that in the meantime, I was already editing my post when you replied (that's why I'd rather have GitHub wait at least 10mn before notifying people of non-urgent comments, I often keep editing my posts for a few minutes). I thought the PR was a fixing PR at first, not a fix for a bug that happened to do what we wanted, hence the confusion.

**Comment by h0lley** [comment]
I am more open now to give the 4.3 behavior a shot.
I have to admit that when working on something like a level scene, it's more useful to continuously launch into that level scene rather than the scene you are currently editing - which often is some game entity that may be part of the level and that you aren't intending to launch as main scene. in that scenario, having to switch tabs back to the level scene before hitting F6 again is the extra step we can do without.

**Comment by domdfcoding** [comment]
Regardless of the behaviour ultimately settled on, I'm not sure the current tooltip text is particularly intuitive. Only the 2nd line changes when there's a scene running (from "Play the edited scene" to "Reload the played scene") but the first line always says "Run Current Scene (F6)" which to me says the scene selected in the editor.  Could the tooltip just be "Play the edited scene (F6)" / "Reload the played scene (F6)" as that better conveys the action at any given moment?

**Comment by Calinou** [comment]
> Could the tooltip just be "Play the edited scene (F6)" / "Reload the played scene (F6)" as that better conveys the action at any given moment?

I've opened a PR for this:

- https://github.com/godotengine/godot/pull/109306
